confmap/converter/expandconverter/expand_test.go

```diff
@@ -159,3 +163,79 @@ func TestNewExpandConverterHostPort(t *testing.T) {
 		})
 	}
 }
+
+func NewTestConverter() (confmap.Converter, *observer.ObservedLogs) {
+	core, logs := observer.New(zapcore.InfoLevel)
+	conv := converter{loggedDeprecations: make(map[string]struct{}), logger: zap.New(core)}
+	return conv, logs
+}
+
+func TestDeprecatedWarning(t *testing.T) {
+	msgTemplate := `Variable substitution using $VAR will be deprecated in favor of ${VAR} and ${env:VAR}, please update $%s`
+	t.Setenv("HOST", "127.0.0.1")
+	t.Setenv("PORT", "4317")
+
+	var testCases = []struct {
+		name             string
+		input            map[string]any
+		expectedOutput   map[string]any
+		expectedWarnings []string
+	}{
+		{
+			name: "no warning",
+			input: map[string]any{
+				"test": "${HOST}:${PORT}",
+			},
+			expectedOutput: map[string]any{
+				"test": "127.0.0.1:4317",
+			},
+			expectedWarnings: []string{},
+		},
+		{
+			name: "one deprecated var",
+			input: map[string]any{
+				"test": "${HOST}:$PORT",
+			},
+			expectedOutput: map[string]any{
+				"test": "127.0.0.1:4317",
+			},
+			expectedWarnings: []string{"PORT"},
+		},
+		{
+			name: "two deprecated vars",
+			input: map[string]any{
+				"test": "$HOST:$PORT",
+			},
+			expectedOutput: map[string]any{
+				"test": "127.0.0.1:4317",
+			},
+			expectedWarnings: []string{"HOST", "PORT"},
+		},
+		{
+			name: "one depracated serveral times",
+			input: map[string]any{
+				"test":  "$HOST,$HOST",
+				"test2": "$HOST",
+			},
+			expectedOutput: map[string]any{
+				"test":  "127.0.0.1,127.0.0.1",
+				"test2": "127.0.0.1",
+			},
+			expectedWarnings: []string{"HOST"},
+		},
```

Comment:
To test the `QuoteMeta` change I commented above, we should add a test like `$HOSTONAME${HOST.NAME}` to check that we are properly escaping characters

Reply:
I have added QuoteMeta because it makes sense, but in practice I'm not really understanding how it is useful or needed, because our regexp will never match `${HOST.NAME}`, and if you attempt to have `$HOST.NAME` you will just end up with `127.0.0.1.NAME` according to my testing. Same if you put other special characters after `$HOST`.

With that in mind, I couldn't think of a good test case, because the `$VAR` way, which is the only one we match, does not play well with special characters anyways.

Reply:
You are right, I mistakenly assumed that `.` would be a valid part of a naked environment variable name for `os.Expand`, but it does not seem like that is the case. 

I guess a test that has both `${HOST.NAME}` and `$HOST_NAME` should be a valid test, right? Taking a look at the implementation of `os.Expand`, it also seems like `$?` would be a valid thing to test, but I doubt anybody is using that in practice :smile: